### PR TITLE
Automate releasing of official packages

### LINKED ISSUE #1333: Automate releasing of proper packages (outside the CI)
Based on https://api3dao.atlassian.net/l/cp/XEq99q90

### PR DESCRIPTION
**Pull request by amarthadan**
Close #1333

Review commit by commit.

### CODE REVIEW

File: docker/scripts/cli.ts [hunk 1]
```diff
@@ -118,10 +118,16 @@ yargs(process.argv.slice(2))
             default: ['latest'],
             type: 'array',
           },
+          'release-branch': {
```
Comment (Siegrift) [inline docker/scripts/cli.ts:146]: 👍 yeah, I felt like this was missing in the previous commits.

File: docker/scripts/utils.ts [hunk 2]
```diff
@@ -1,14 +1,28 @@
...
-import omit from 'lodash/omit';
+import has from 'lodash/has';
 import { logger } from '@api3/airnode-utilities';
 import { goSync } from '@api3/promise-utils';
 
+const hideSecrets = (command: string, options?: ExecSyncOptions) => {
+  // There are currently 2 secrets being logged:
+  // * DockerHub auth token - stdin
+  // * GitHub auth token - remote URL
+  // Third secret, the NPM auth token, is stored in a file therefor not logged
```
Comment (Siegrift) [inline docker/scripts/utils.ts:11]: therefr**E**

File: docker/scripts/utils.ts [hunk 3]
```diff
@@ -1,14 +1,28 @@
...
+
+  // DockerHub auth token
+  const hiddenOptions = has(options, 'input') ? { ...options, input: '***' } : options;
+
+  // GitHub auth token
+  const hiddenCommand = command.replace(/(https:\/\/x-access-token):(.*?)(@github\.com.*$)/, '$1:***$3');
```
Comment (Siegrift) [inline docker/scripts/utils.ts:17]: Btw. is the token really logged? I remember when I was doing some CI stuff on my personal projects that GitHub was able to star ("****") the secrets. Even if they were output as part of the message. Doesn't that happen here?

**Review by Siegrift — COMMENTED** [review]
Btw. have you also looked at https://github.com/changesets/action ? It does the PR opening and potentially NPM package releasing. I'm sorry for not giving the link sooner, but it just now came to my mind that I saw this thing being mentioned in the changeset docs...

I am fine with our own implementation though...

**Comment by Siegrift** [comment]
Did a shallow pass only today, will take a proper look tomorrow.

**Comment by amarthadan** [comment]
> Btw. have you also looked at https://github.com/changesets/action ? It does the PR opening and potentially NPM package releasing. I'm sorry for not giving the link sooner, but it just now came to my mind that I saw this thing being mentioned in the changeset docs...
> 
> I am fine with our own implementation though...

No worries, I knew about it :slightly_smiling_face: I decided against it mainly because it's just a CI action and I wanted a solution you can run locally as well. I could have used some of the code though...

File: docker/scripts/npm.ts [hunk 4]
```diff
@@ -124,14 +141,68 @@ export const publishPackages = async (npmRegistry: string, npmTag: string, snaps
     throw new Error('Missing NPM authentication token');
   }
 
+  return npmAuthToken;
+};
+
+const prePublishSetup = async (npmRegistry: string, releaseBranch?: string) => {
+  const npmRegistryUrl = getNpmRegistryUrl(npmRegistry);
+  fetchProject(!releaseBranch);
```
Comment (Siegrift) [inline docker/scripts/npm.ts:149]: Nit: fetch suggests that it's asynchronous

File: docker/scripts/npm.ts [hunk 5]
```diff
@@ -124,14 +141,68 @@ export const publishPackages = async (npmRegistry: string, npmTag: string, snaps
...
+      // https://github.com/yarnpkg/yarn/issues/7823#issuecomment-737248277
+      runCommand(`npm dist-tag add ${npmPackage}@${releaseVersion!} ${additionalNpmTag}`);
+    }
   }
-  // Ignoring commands' outputs because of the weird text colorization.
-  runCommand(`yarn changeset version --snapshot ${npmTag}`, { cwd: '/build', stdio: 'ignore' });
-  runCommand(`yarn changeset publish --no-git-tag --snapshot --tag ${npmTag}`, { cwd: '/build', stdio: 'ignore' });
+
+  git.setIdentity('API3 Automation', 'automation@api3.org');
+  github.authenticate();
```
Comment (Siegrift) [inline docker/scripts/npm.ts:204]: It's strange that this is sync. Maybe rename `authenticateSync`?

**Review by Siegrift — APPROVED** [review]
👍 LGTM,

I'd love to see this in action though. Maybe we could do a test release?

**Comment by amarthadan** [comment]
> I'd love to see this in action though. Maybe we could do a test release?

I'd leave it for v0.10. I don't really want to mess up the official API3 registry with random versions.

**Pull request merged**